pkg/client/logs.go

```diff
@@ -19,59 +19,211 @@ package client
 import (
 	"fmt"
 	"io"
-	"os"
-	"strings"
+	"sync"
 
-	"github.com/heptio/sonobuoy/pkg/config"
-	"github.com/sirupsen/logrus"
+	"github.com/pkg/errors"
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 )
 
-var (
-	podLogSeparator = strings.Repeat("-", 79)
+const (
+	bufSize = 4096
 )
 
-// Logs gathers the logs for the containers in the sonobuoy namespace and prints them
+// Reader provides an io.Reader interface to a channel of bytes.
+// The first error received on the error channel will be returned by Read and all subsequent calls to Read.
+type Reader struct {
+	bytestream chan []byte
+	errc       chan error
+	done       chan struct{}
 
-func (c *SonobuoyClient) GetLogs(cfg *LogConfig) error {
-	// TODO(EKF): Stream to a writer instead of just stdout
-	if *cfg.Follow {
-		return streamLogs(c.Client, cfg.Namespace, config.MasterPodName, &v1.PodLogOptions{Follow: true})
+	// Used for when one message is too large for the input buffer
+	overflowBuffer []byte
+	err            error
+}
+
+// NewReader returns a configured Reader.
+func NewReader(bytestream chan []byte, errc chan error) *Reader {
+	reader := &Reader{
+		bytestream:     bytestream,
+		errc:           errc,
+		overflowBuffer: []byte{},
+		err:            nil,
+	}
+	return reader
+}
+
+// Read tries to fill up the passed in byte slice with messages from the channel.
+// Read manages the message overflow ensuring no bytes are missed.
+// If an error is set on the reader it will return the error immediately.
+func (r *Reader) Read(p []byte) (int, error) {
+	// Always return the error if it is set.
+	if r.err != nil {
+		return 0, r.err
+	}
+
+	// n is the number of bytes written to `p`.
+	n := 0
+
+	// Send any overflow before grabbing new messages.
+	if len(r.overflowBuffer) > 0 {
+		// If we need to chunk it, copy as much as we can and reduce the overflow buffer.
+		if len(r.overflowBuffer) > len(p) {
+			copy(p, r.overflowBuffer[:len(p)])
+			r.overflowBuffer = r.overflowBuffer[len(p):]
+			return len(p), nil
+		}
+		// At this point the entire overflow will fit into the buffer.
+		copy(p, r.overflowBuffer)
+		n += len(r.overflowBuffer)
+		r.overflowBuffer = []byte{}
+		// This is suboptimal (does not use entire buffer) but gets output to the user faster.
+		return n, nil
 	}
 
-	pods, err := c.Client.CoreV1().Pods(cfg.Namespace).List(metav1.ListOptions{})
+	// One message a time.
+	select {
+	// it's ok to keep reading messages and miss an error.
```

Comment:
what does this mean? will we keep returning this error if it's called repeatedly?

Reply:
this is acknowledging that if both channels are ready, select will randomly choose either errc or bytestream. if it decides to choose bytestream we'll keep reading even though we've encountered an error somewhere, but it's ok to keep reading, eventually select will give us our error and it will get propagated as expected.